### PR TITLE
IOColor (IOContext part 2?)

### PR DESCRIPTION
**Pull request by vtjnash**
While working on the original IOContext PR, I had created a PoC of fully handling nested color. Here, I'm working on finally reviving and finishing that, building on all of the other new work towards better color handling.

Some color / format nesting in action:
```julia
julia> let buf = IOBuffer()
           Base.with_output_color(:red, IOContext(buf, :color => true)) do io; print(io, "red",
               Base.text_colors[:underline], " underline") # emit some formatting code directly too
           Base.with_output_color(:blue, io) do io; print(io, " blue") # nested blue
           Base.with_output_color(:bold, io) do io; print(io, " bold") # nested bold
           Base.with_output_color(:green, io) do io; print(io, " green\ngreen") # with a newline!
           end; print(io, " bold",
               Base.text_colors[:underline], " underline")
           end; print(io, " blue")
           end; print(io, " red")
           end
           String(take!(buf))
       end
"\e[31mred\e[4m underline\e[34m blue\e[1m bold\e[32m green\e[22m\e[39m\n\e[32m\e[1mgreen\e[34m bold\e[4m underline\e[22m blue\e[31m red\e[39m"
```
image

There's also some demo code in here of how that output could also be turned into html (ref https://github.com/JuliaLang/julia/issues/5239), which was probably the real purpose behind this work (although not actually finished or enabled here).

### CODE REVIEW

File: base/formatio.jl [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        buf = IOBuffer(read=false, write=true, append=true)
+        buf.seekable = false
+        return new(buf, Symbol[], Int[], Int[])
+    end
+end
+
+pipe_writer(io::IOFormatBuffer) = io.buf
+pipe_reader(io::IOFormatBuffer) = error("IOFormatBuffer not readable")
+
+function with_output_color(f::Function, color::Union{Int, Symbol}, io::IO, args...; bold::Bool = false)
```
Comment (ararslan) [inline base/formatio.jl]: Perhaps a good opportunity to rename `with_output_color` to something less ugly?

Comment (StefanKarpinski) [reply base/formatio.jl]: I had some recollection that we'd already renamed it to `styled` or something like that?

Comment (ararslan) [reply base/formatio.jl]: We changed `print_with_color` to `printstyled`, but left `with_output_color` unchanged, I forget why.

Comment (ararslan) [reply base/formatio.jl]: Maybe `withstyle(f, io, color=..., bold=...)`?

Comment (ararslan) [reply base/formatio.jl]: Or perhaps just `styled(f, io, color=, bold=)`?

**Comment by vtjnash** [comment]
Anyone seen this error before?
https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.27586/job/9s6uqvwkwn1ebl8q#L3277
```
The global RNG seed was 0xd11701fbc3bda26f50d41cb8a2efb875.
3278
3279Error in testset SuiteSparse:
3280Error During Test at C:\projects\julia\julia-6ae2e7d268\share\julia\stdlib\v0.7\SuiteSparse\test\umfpack.jl:92
3281  Got exception LinearAlgebra.SingularException(0) outside of a @test
3282  LinearAlgebra.SingularException(0)
3283  Stacktrace:
3284   [1] #lu#1(::Bool, ::Function, ::SparseArrays.SparseMatrixCSC{Float64,Int32}) at C:\projects\julia\usr\share\julia\stdlib\v0.7\SuiteSparse\src\umfpack.jl:159
3285   [2] lu(::SparseArrays.SparseMatrixCSC{Float64,Int32}) at C:\projects\julia\usr\share\julia\stdlib\v0.7\SuiteSparse\src\umfpack.jl:152
3286   [3] macro expansion at C:\projects\julia\julia-6ae2e7d268\share\julia\stdlib\v0.7\SuiteSparse\test\umfpack.jl:76 [inlined]
3287   [4] macro expansion at C:\projects\julia\usr\share\julia\stdlib\v0.7\Test\src\Test.jl:1153 [inlined]
3288   [5] macro expansion at C:\projects\julia\julia-6ae2e7d268\share\julia\stdlib\v0.7\SuiteSparse\test\umfpack.jl:17 [inlined]
3289   [6] macro expansion at C:\projects\julia\usr\share\julia\stdlib\v0.7\Test\src\Test.jl:1080 [inlined]
3290   [7] top-level scope at C:\projects\julia\julia-6ae2e7d268\share\julia\stdlib\v0.7\SuiteSparse\test\umfpack.jl:7 [inlined]
```

File: base/formatio.jl (hunk 1 above)
Comment (vtjnash) [reply base/formatio.jl]: `bold` being some arbitrary special kwarg should probably be deprecated

Comment (ararslan) [reply base/formatio.jl]: Why's that? In `printstyled` we have `color` and `bold` keyword arguments and that seems to work quite nicely.

**Comment by KristofferC** [comment]
I think this can be closed since my guess is that this will be incompatible with the styled strings stuff.

**Comment by vtjnash** [comment]
It is mostly all compatible, as they have similar designs, but a lot of work possibly to filter through here and pick up any fixes that didn't already get separately put in master

**Pull request closed without merging**